source/client/ClientState.py

```diff
@@ -57,31 +65,81 @@ def reset(self):
-                self.played_cards.setdefault(key, []).append(card)
-    
+        # Check ruleset to determine whether self.played_cards = all visible cards or cards that this client played.
+        if not self.rules.Shared_Board:
+            self.rules.canPlay(prepared_cards, self.played_cards, self.round)
+            for key, card_group in prepared_cards.items():
+                for card in card_group:
+                    self.hand_cards.remove(card)
+                    self.played_cards.setdefault(key, []).append(card)
+        elif self.rules.Shared_Board:
+            # Review Notes
+            # todo: move these to documentation
+            #  unlike in HandAndFoot, where self.played_cards was used to check rules.
+            # in Liverpool and other shared board games need to consider all of the played cards.
+            # Played cards (in deserialized form) are in visible_scards (in serialized form), which is obtained
+            # from controller.
+            # (Path taken by visible_scards:
+            #          Tableview gets the serialized cards every msec to keep display up to date,
+            #          In handview.update tableview.visible_scards list is passed to handview.visible_scards
+            #          No need to process this unless playing cards, in which case visible_scards passed
+            #          to controller and then to clientState, where only list item is deserialized and put in
+            #          dictionary self.played_cards
+            self.played_cards = self.rules.restoreRunAssignment(visible_scards[0], self.round)
+            # restoreRunAssignment converts all serialized cards to cards and processes self.played_cards
+            # that are in runs so that positions of Wilds and Aces are maintained.
+            # Review Note:  This is put in rules because I can imagine other shared board games
+            # will have different rules, such as whether Aces can be both high and low.
+            #
+            # todo:
+            #  tried to get canPlay to return combined dictionary, but then got error that indicated line below was
+            # converting self.played_cards into a boolean (?). Commented out next line and changed return combined_cards
+            # to return True in rules.canPlay.  This means need to process runs again.
+            # If no exception raised, wanted canPlay to return cards in proper order, ready to be transmitted to server.
+            # todo: debug  next line.
+            # self.played_cards  = self.rules.canPlay(prepared_cards, self.played_cards, self.player_index, self.round)
+            self.rules.canPlay(prepared_cards, self.played_cards, self.player_index, self.round)
+
+            combined_cards = self.rules.combineCardDicts(self.played_cards, prepared_cards)
+            self.played_cards = {}
+            for k_group, card_group in combined_cards.items():
+                if k_group[1] >= self.rules.Meld_Threshold[self.round][0]:
```

Comment:
This comparison works because you put all the sets first when you generate the integer keys for the k_group tuple (I assume not a list, I haven't gotten quite that far). You need a comment where you generate those indices that that is what you are doing so it is documented in code, Ideally with the phrase k_group in the comment so it can be found. Its elegant but needs in code documentation.